### PR TITLE
Ryujinx game provider addon

### PR DESCRIPTION
**Pull request by albfan**
### Description
Game provider for the Ryujinx emulator

- TODO:
   - Depends on a PR on Ryujinx to list available games https://github.com/Ryujinx/Ryujinx/pull/2131
   - Listen to ESC to end runned game
   - Setup Ryujinx path
   - Run on same screen as kodi (for multimonitor setup)
   - Use `bmcplugin.setResolvedUrl` isntead for subprocess, right now I get (file cannot be played with any known emulator)
 
- Good to have:
   - Integration with "Add game"
   - Integration with gamepad controller (for joycons)
  
### Checklist:
- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-plugins/blob/master/CONTRIBUTING.md) document
- [x] Each add-on submission should be a single commit with using the following style: [plugin.video.foo] v1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practice but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-plugins/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.
- If you see no activity on your PR after a week (so at least one weekend has passed) then please go to the #kodi-dev freenode IRC channel to reach out to the team

### CODE REVIEW

**Comment by basrieter** [comment]
This seems to be an incomplete PR (with TODO's) on the `master` branch. Please only create complete PR's against the correct target branch: `leia` or `matrix`.

**Comment by albfan** [comment]
@basrieter thanks. Main TODO is to get custom path for Ryujinx if needed (but Ryujinx should be found on path usually)

I try to set this as draft but looks I have no rights to do it.

Other TODOs are questions for kodi developers:

- Can I run external emulator with kodi?
- Can I get the kodi screen in a multimonitor display to start emulator on same screen?
- Can I listen to escape or exit on Kodi to kill external app?

**Comment by basrieter** [comment]
I would suggest you do this in your own repository and create a pull request when it is complete. For your add-on development questions I would redirect you to the Kodi forums.

**Comment by albfan** [comment]
Sure

**Comment by albfan** [comment]
https://github.com/Ryujinx/Ryujinx/pull/2310 will help in running emulator headless. Still not sure about stopping it

**Comment by albfan** [comment]
From this info: https://kodi.wiki/view/Releases matrix is the last release 19.09, so rebasing on top of that

**Comment by albfan** [comment]
Refreshing TODO list. Will ask in forums as suggested by @basrieter 

TODO:
    - Depends on a PR on Ryujinx to list available games Integration with kodi Ryujinx/Ryujinx#2131
    - Listen to ESC to end runned game
    - Setup Ryujinx path with config
    - Run on same screen as kodi (for multimonitor setup) can be defined with config till then
    - Use bmcplugin.setResolvedUrl isntead for subprocess, right now I get (file cannot be played with any known emulator)

- Good to have:
    - Integration with "Add game"
    - Integration with gamepad controller (for joycons)

**Comment by albfan** [comment]
Questions posted on forum: https://forum.kodi.tv/showthread.php?tid=364906

File: .github/ISSUE_TEMPLATE/broken_addon.md
Comment (basrieter) [inline .github/ISSUE_TEMPLATE/broken_addon.md]: Don't include in submission.

File: .github/ISSUE_TEMPLATE/config.yml
Comment (basrieter) [inline .github/ISSUE_TEMPLATE/config.yml]: Don't include in submission.

File: README.md
Comment (basrieter) [inline README.md]: Don't include in submission.

File: PULL_REQUEST_TEMPLATE.md
Comment (basrieter) [inline PULL_REQUEST_TEMPLATE.md]: Don't include in submission.

File: ISSUE_TEMPLATE.md
Comment (basrieter) [inline ISSUE_TEMPLATE.md]: Don't include in submission.

File: .github/ISSUE_TEMPLATE/violating_addon.md
Comment (basrieter) [inline .github/ISSUE_TEMPLATE/violating_addon.md]: Don't include in submission.

**Review by basrieter — CHANGES_REQUESTED** [review]
You include a bunch of standard files from our repo. Please make sure you rebase before you do a PR.

File: .github/ISSUE_TEMPLATE/broken_addon.md
Comment (albfan) [reply .github/ISSUE_TEMPLATE/broken_addon.md]: I notice on rebase, and forget to push changes. Solved

File: plugin.ryujix/addon.xml
Comment (enen92) [inline plugin.ryujix/addon.xml]: addon id doesn't match folder name. Please make it consistent.
We use (type).(content).(name), eg: plugin.program.ryujinx

File: plugin.ryujix/addon.xml
Comment (enen92) [inline plugin.ryujix/addon.xml]: en_GB

File: plugin.ryujix/addon.xml
Comment (enen92) [inline plugin.ryujix/addon.xml]: Missing licence tag and source tag (see: https://kodi.wiki/view/Addon.xml)

File: plugin.ryujix/main.py
Comment (enen92) [inline plugin.ryujix/main.py]: better move this to a configurable path in settings.xml. Request the user to fill it on the first run if not valid.

File: plugin.ryujix/main.py
Comment (enen92) [inline plugin.ryujix/main.py]: xbmcvfs.translatePath (xbmc.translatePath was already dropped in nexus)

File: plugin.ryujix/main.py
Comment (enen92) [inline plugin.ryujix/main.py]: better to use os.path.join

**Review by enen92 — COMMENTED** [review]
A few comments. Thanks for the submission.
I am a big fan of your miraclecast work btw :)

File: plugin.ryujix/addon.xml
Comment (albfan) [reply plugin.ryujix/addon.xml]: got it

File: plugin.ryujix/main.py
Comment (albfan) [reply plugin.ryujix/main.py]: Sure, I'm looking for an example to connect settings for add on and be able to setup this first (Any good example out there?)

Comment (enen92) [reply plugin.ryujix/main.py]: See here an example (we have two types of addon settings, this one is the new system) of a string edit control for a string setting: https://github.com/xbmc/repo-plugins/blob/matrix/plugin.whereareyou/resources/settings.xml#L21-L35

You basically need a settings.xml file under addonid/resources folder. Full documentation on the format can be found here: https://kodi.wiki/view/Add-on_settings_conversion

To get it from python [you can use](https://codedocs.xyz/xbmc/xbmc/group__python__xbmcaddon.html#ga722c9c4fdc6a320e32d448bd180af650) `xbmcaddon.Addon().getSettingString(id)` where id in this case is `mappings`. 

To open the settings dialog from python you can issue: `xbmcaddon.Addon().openSettings()`

Comment (albfan) [reply plugin.ryujix/main.py]: Cool working on it. I found this too: [HOW-TO write GUI settings for XBMC python plugins and scripts (addons)](https://forum.kodi.tv/showthread.php?tid=29577)

Comment (enen92) [reply plugin.ryujix/main.py]: Yeah that one is based on the old settings format which is basically equivalent to this wiki page: https://kodi.wiki/view/Add-on_settings. However as stated in the first line/disclaimer this format is somehow deprecated from matrix and up, will probably be dropped in some future version

**Comment by enen92** [comment]
Please take a look at CI failures, https://github.com/xbmc/repo-plugins/pull/3510/checks?check_run_id=3918149094
The automated checks can figure out a few issues automatically :)

File: plugin.ryujix/main.py
Comment (albfan) [reply plugin.ryujix/main.py]: Solved!

**Comment by albfan** [comment]
Waiting for the CI

Still pending.

- How to warn user if settings are not setup? If emulator is not set I prefer something better than error (check log file)
- I modify emulator to listen to joycon home button (guide button) to launch stop game dialog but I miss be able from kodi to listen to exit on HDMI remote TV control (like it does for a movie) @enen92 @basrieter any ideas? I suppose implement a game emulator addon instead of a program will make that possible

I try something like:

```
diff --git i/plugin.program.ryujinx/addon.xml w/plugin.program.ryujinx/addon.xml
index 032a45d0c..739335eda 100644
--- i/plugin.program.ryujinx/addon.xml
+++ w/plugin.program.ryujinx/addon.xml
@@ -10,6 +10,17 @@
     <extension point="xbmc.python.pluginsource" library="main.py">
         <provides>executable game</provides>
     </extension>
+    <extension point="kodi.gameclient"
+      library_android="game.moonlight.so"
+      library_linux="game.moonlight.so"
+      library_osx="game.moonlight.dylib"
+      library_windx="game.moonlight.dll"
+      library_wingl="game.moonlight.dll">
+      <platforms></platforms>
+      <extensions>nsp</extensions>
+      <supports_vfs>true</supports_vfs>
+      <supports_no_game>true</supports_no_game>
+    </extension>
     <extension point="xbmc.addon.metadata">
         <summary lang="en_GB">Ryujinx emulator on Kodi</summary>
         <description lang="en_GB">Ryujinx frontend for Kodi.</description>
```

But cannot run an nsp game that way (did I miss executable path?)

- Can I ask kodi which monitor is it using in a multimonitor setup?

**Comment by albfan** [comment]
@basrieter @enen92 I map home button of joycons to stop emulation, so if exit from remote tv cannot be used to stop emulation, that is mostly solved.

Wonder if it is possible to use joycons to interact with kodi, I see some documentation  https://kodi.wiki/view/HOW-TO:Configure_controllers but not sure how to create a new game controller.

Is it detected by vendor:product?

**Comment by enen92** [comment]
@albfan buttonmaps are out of the scope of python plugins. Never did it myself but I think you have to create a buttonmap for your joycon controller (see https://kodi.wiki/view/HOW-TO:Share_button_maps). 
There's no way to grab the button control from python plugins. You either map a key to a specific action (using the keymapper add-on) or you create a script instead of a plugin on which you design your own window and listen for actions on the window (https://codedocs.xyz/xbmc/xbmc/group__python__xbmcgui__window__cb.html#ga9ab0edbacfaba93c64c1f2ec6f335884).

**Comment by enen92** [comment]
To check if settings are filled you can check the value of the setting on the add-on entry point:

my_path = xbmcaddon.Addon().getSetting("myPathsettingId")
if not my_path or not xbmcvfs.exists(xbmcvfs.translatePath(my_path):
   xbmcgui.Dialog.ok("my_addon", "setting not configured")
   xbmcaddon.Addon().openSettings()

**Comment by albfan** [comment]
Many thanks!

I will implement those details and fill the README with exact instructions (to install emulator) A fork until integration is merged on Ryujinx

**Comment by ezamelczyk** [comment]
so this is dead right?

**Comment by albfan** [comment]
No progress so far, yes.

**Pull request closed without merging**
